### PR TITLE
Add active learning example notebook

### PR DESCRIPTION
**Pull request by huiwengoh**
Adds an example notebook demonstrating sample usage of the active learning functions in `cleanlab.multiannotator`

note: https://github.com/cleanlab/cleanlab/pull/538 has to be approved and merged before merging this example notebook

### CODE REVIEW

File: active_learning_multiannotator/active_learning.ipynb [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Train model and obtain predicted probabilites\n",
+    "\n",
+    "First, we train our model on a set of consensus labels obtained using majority vote to get the out-of-sample predicted probabilities for both the labeled and unlabeled data. \n",
+    "\n",
+    "The function will return two sets of predicted probabilites, `pred_probs_labeled` are the probabilites for examples that have existing annotator labels (they correspond directly with the features `X_labeled`), whereas `active_learning_score_unlabeled` are the probabilites for examples that do not have any annotator labels (they correspond directly with the features `X_unlabeled`). These predicted probabilities will later be used to compute the active learning score.\n",
```
Comment (jwmueller) [inline active_learning_multiannotator/active_learning.ipynb]: In general, would always first use the full terminology: 
"predicted class probabilities" the first time this concept is mentioned in a notebook.

And later "predicted probabilities". For the shorthand version, I would use term "predictions" over "probabilities" (as long as context makes it clear these are probabilistic predictions)

File: active_learning_multiannotator/active_learning.ipynb [hunk 2]
```diff
@@ -0,0 +1,419 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Using Cleanlab to Determine Examples to Collect Additional Labels for in an Active Learning Loop to Improve Classification Models "
```
Comment (jwmueller) [inline active_learning_multiannotator/active_learning.ipynb]: We want to be very clear to distinguish this from the traditional academic Active Learning. What is different in our setting is there are multiple annotators, and we can have annotators review already labeled examples in order to collect a higher quality overall set of labels.

File: active_learning_multiannotator/active_learning.ipynb [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Train model and obtain predicted probabilites\n",
+    "\n",
+    "First, we train our model on a set of consensus labels obtained using majority vote to get the out-of-sample predicted probabilities for both the labeled and unlabeled data. \n",
+    "\n",
+    "The function will return two sets of predicted probabilites, `pred_probs_labeled` are the probabilites for examples that have existing annotator labels (they correspond directly with the features `X_labeled`), whereas `active_learning_score_unlabeled` are the probabilites for examples that do not have any annotator labels (they correspond directly with the features `X_unlabeled`). These predicted probabilities will later be used to compute the active learning score.\n",
+    "\n",
+    "Note that our `train_model` function here also computes the model accuracy however if you have no test data, you can set `X_test=None` and `labels_test=None` and the function will still compute the predicted probabilites."
```
Comment (jwmueller) [inline active_learning_multiannotator/active_learning.ipynb]: `train_model` should perhaps be called `fit_predict` like sklearn terminology? (don't recall exact sklearn terminology so you could look into the most similar functions they have)

Comment (jwmueller) [reply active_learning_multiannotator/active_learning.ipynb]: clarify this is supposed to be a user-defined function, ie. 

You should replace this function with your own code for model training/inference on a given dataset

File: active_learning_multiannotator/active_learning.ipynb [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+    "plt.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "From the plot above, we can see that the model accuracy increases steadily with each additional round of collecting more labels, getting improved consensus labels, and model training."
+   ]
+  }
```
Comment (jwmueller) [inline active_learning_multiannotator/active_learning.ipynb:693]: also show how to get annotator quality scores at the end of the notebook?

File: active_learning_multiannotator/active_learning.ipynb [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+   "source": [
+    "import numpy as np\n",
+    "import pandas as pd\n",
+    "from sklearn.ensemble import ExtraTreesClassifier\n",
+    "import matplotlib.pyplot as plt\n",
+    "\n",
+    "from cleanlab.multiannotator import get_majority_vote_label, get_label_quality_multiannotator, get_active_learning_scores\n",
+    "\n",
+    "from utils.model_training import train_model\n",
+    "from utils.active_learning import get_idx_to_relabel, setup_next_iter_data, add_new_annotator\n"
```
Comment (jwmueller) [inline active_learning_multiannotator/active_learning.ipynb]: `get_idx_torelabel()` seems like crucial function and should probably be defined in this notebook to be explicit.

Comment (jwmueller) [reply active_learning_multiannotator/active_learning.ipynb]: also should we call it "to_relabel" or "to_label".

We are collecting a new label for these datapoints, whereas "relabel" sounds to me like overwriting the existing label

File: active_learning_multiannotator/active_learning.ipynb [hunk 6]
```diff
@@ -0,0 +1,419 @@
...
+    "The code cell below shows how to combine the labeled and unlabeled examples before ranking them and obtaining their respective indices to collect more labels."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "num_examples_to_label = 100 # you can pick how many examples you want to collect more labels for\n",
```
Comment (jwmueller) [inline active_learning_multiannotator/active_learning.ipynb]: keyword to use here is "batch"

File: active_learning_multiannotator/active_learning.ipynb [hunk 7]
```diff
@@ -0,0 +1,419 @@
...
+    "\n",
+    "# train model to get out-of-sample predicted probabilies for both labels and unlabeled data\n",
+    "pred_probs_labeled, pred_probs_unlabeled, model_accuacy = train_model(\n",
+    "    ExtraTreesClassifier(),\n",
+    "    X_labeled,\n",
+    "    consensus_labels,\n",
+    "    cv_n_folds=5,\n",
+    "    X_unlabeled=X_unlabeled,\n",
+    "    X_test=X_test,\n",
+    "    labels_test=true_labels_test,\n",
```
Comment (jwmueller) [inline active_learning_multiannotator/active_learning.ipynb]: we should separate out stuff user is supposed to implement vs optional stuff. 
So `train_model()` should not calculate the model accuracy each round. That should be done by another function `evaluate_accuracy()` which takes `X_test` and `labels_test` and trained-model and outputs the accuracy of the model on X_test.  This function should clarify that `labels_test` should be high-quality gold labels that can be trusted to not have errors.

File: active_learning_multiannotator/active_learning.ipynb [hunk 8]
```diff
@@ -0,0 +1,419 @@
...
+    "    (\n",
+    "        multiannotator_labels, relabel_idx_combined, X_labeled, X_unlabeled, pred_probs_labeled, \n",
+    "        pred_probs_unlabeled, extra_labels_labeled, extra_labels_unlabeled,\n",
+    "    ) = setup_next_iter_data(\n",
+    "        multiannotator_labels, relabel_idx, relabel_idx_unlabeled, X_labeled, X_unlabeled, pred_probs, \n",
+    "        pred_probs_unlabeled, extra_labels_labeled, extra_labels_unlabeled,\n",
+    "    )\n",
+    "\n",
+    "    # add a new annotator that provides new labels for the examples with lowest active learning scores (indices obtained above)\n",
+    "    multiannotator_labels = add_new_annotator(\n",
```
Comment (jwmueller) [inline active_learning_multiannotator/active_learning.ipynb:385]: this makes it sound like the new batch of data must be annotated by a single new annotator, so try to think of how to make it clear that new batch of data can be labeled by multiple annotators (existing annotators could handle some of these examples as well)

File: active_learning_multiannotator/active_learning.ipynb (hunk 7 above)
Comment (jwmueller) [reply active_learning_multiannotator/active_learning.ipynb]: also should clarify evaluating the model accuracy in each round is an optional step

File: active_learning_multiannotator/utils/active_learning.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+
+
+def setup_next_iter_data(
+    multiannotator_labels,
+    relabel_idx,
+    relabel_idx_unlabeled,
+    X,
+    X_unlabeled,
+    pred_probs,
+    pred_probs_unlabeled,
```
Comment (jwmueller) [inline active_learning_multiannotator/utils/active_learning.py:41]: does this function actually need to deal with pred_probs at all? or could we just ignore pred_probs since they'll be recomputed next round?

File: active_learning_multiannotator/active_learning.ipynb (hunk 1 above)
Comment (cgnorthcutt) [reply active_learning_multiannotator/active_learning.ipynb]: IMO anything `pred_proba` related should be called predicted probabilities. Anything `predict` should be called predictions. predictions = argmax(predicted probs), so probably best to use predicted probabilities when referring to probabilities and predictions when referring to labels that are predicted. Will chat with jonas separately and find a consistent nomenclature.

Comment (jwmueller) [reply active_learning_multiannotator/active_learning.ipynb]: Sounds good to me, we can just always stick with shorthand term "predicted probabilities" like most of our existing docstrings and not go shorter than that

File: active_learning_multiannotator/active_learning.ipynb (hunk 5 above)
Comment (huiwengoh) [reply active_learning_multiannotator/active_learning.ipynb]: `idx_to_label` makes sense! Agree that `relabel` could be confusing.

File: active_learning_multiannotator/active_learning.ipynb (hunk 4 above)
Comment (huiwengoh) [reply active_learning_multiannotator/active_learning.ipynb:693]: Added a section at the end of notebook that demonstrates basic crowdlab functionality!

File: active_learning_multiannotator/active_learning.ipynb (hunk 2 above)
Comment (huiwengoh) [reply active_learning_multiannotator/active_learning.ipynb]: I edited the title slightly, but I'm not super certain on how to make it super clear and concise - any suggestions would be helpful!

File: active_learning_multiannotator/utils/model_training.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+
+def fit_predict_proba(
+    model,
+    X,
+    labels,
+    cv_n_folds=5,
+    X_unlabeled=None,
+):
+    if X_unlabeled is None:
```
Comment (jwmueller) [inline active_learning_multiannotator/utils/model_training.py:30]:
```suggestion
   """ In your applications, replace this with your code to train model on (X, labels) and produce held-out predictions for X and X_unlabeled. Held-out predictions for X can be produced by training via cross-validation. """
    if X_unlabeled is None:
```

File: active_learning_multiannotator/utils/active_learning.py [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+    extra_labels=None,
+    extra_labels_unlabeled=None,
+):
+
```
Comment (jwmueller) [inline active_learning_multiannotator/utils/active_learning.py:22]:
```suggestion
    """ Updates all of these inputs after additional labels have been collected in a single active learning round. """
```

Comment (jwmueller) [reply active_learning_multiannotator/utils/active_learning.py:22]: Feel free to implement a better suggestion, mine is pretty crude above

File: active_learning_multiannotator/utils/active_learning.py [hunk 12]
```diff
@@ -0,0 +1,82 @@
+import numpy as np
```
Comment (jwmueller) [inline active_learning_multiannotator/utils/active_learning.py:5]: would be nice to have one-liner comment at top describing the purpose of this file

File: active_learning_multiannotator/utils/active_learning.py [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+        X_unlabeled,
+        pred_probs,
+        pred_probs_unlabeled,
+        extra_labels,
+        extra_labels_unlabeled,
+    )
+
+
+def add_new_annotator(multiannotator_labels, extra_labels, relabel_idx):
+    def get_random_label(annotator_labels):
```
Comment (jwmueller) [inline active_learning_multiannotator/utils/active_learning.py:76]:
```suggestion
    def get_random_label(annotator_labels):
        """ Collects more labels for the examples whose indices are specified in `relabel_idx` from a single new annotator, and adds the new labels to dataset. In your applications, these new labels could instead  come from multiple existing annotators, just be sure to update the `multiannotator_labels` appropriately.
```

File: active_learning_multiannotator/active_learning.ipynb [hunk 14]
```diff
@@ -0,0 +1,677 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Using Cleanlab to Determine Examples to Collect Additional Labels for in Data Labeled by Multiple Annotators to Improve Classification Models in Active Learning"
```
Comment (jwmueller) [inline active_learning_multiannotator/active_learning.ipynb]:
```suggestion
    "# Active Learning with Multiple Data Annotators via CROWDLAB"
```

File: active_learning_multiannotator/active_learning.ipynb [hunk 15]
```diff
@@ -0,0 +1,677 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This example demonstrates how to compute an active learning score for every datapoint in your dataset, and shows how a classification model improves after relabeling examples for multiple rounds. The examples in the dataset will have been annotated by multiple annotators, and not all examples have to be labeled at the beginning. The computed active learning score will indicate if you should collect more labels for an already-labeled example (thus having a larger number of annotators label a potentially difficult example), or get a label for an example that has not been labeled.\n",
```
Comment (jwmueller) [inline active_learning_multiannotator/active_learning.ipynb]:
```suggestion
    "This notebook demonstrates a practical approach to active learning for training  classification models with cleanlab. In active learning, we aim to construct a labeled dataset by collecting the fewest labels that still allow us to train an accurate classifier model. Here we assume data labeling is done in **batches**, and between these data labeling rounds, we retrain our classifier and combine it with CROWDLAB to decide what examples (i.e. datapoints) to label next round. We assume there are **multiple** annotators to label the data, but since their labels are not necessarily perfect, we sometimes wish to ask a new annotator to provide an extra label for a previously labeled example. This active learning setup is more widely applicable than academic studies which only consider labeling one example in each round or being limited to a single annotation per example.\n",
"\n",
 "cleanlab provides two useful estimates for multi-annotator active learning: (1) an active learning score quantifying how desirable it is to collect an additional label for every possible example, (2) accurate consensus labels for the previously labeled examples  that can be used for classifier model training.\n",
 "\n",
 "This notebook demonstrates how to compute these easily for use in sequential active learning, showing how a classification model iteratively improves after (re)labeling examples for multiple rounds. Some examples in the dataset will have been labeled by multiple annotators, and not all examples have to have been labeled at the beginning. cleanlab's active learning score indicates if you should collect more labels for an already-labeled example (thus having a larger number of annotators label a potentially difficult example), or get a label for a potentially informative example that has not yet been labeled.\n",
```

File: active_learning_multiannotator/active_learning.ipynb [hunk 16]
```diff
@@ -0,0 +1,677 @@
...
+    "\n",
+    "This notebook will demonstrate the following steps:\n",
+    "\n",
+    "1. Train a model on the current consensus labels to obtain out-of-sample predicted probabilities for each datapoint\n",
+    "2. Compute active learning scores for every datapoint, this score indicates the quality and confidence of each datapoint's label, where a low active learning score indicates that we should get more\n",
+    "3. Collect additional labels for the examples with the lowest active learning scores\n",
+    "4. Repeat the steps above with a new set of consensus labels obtained after collecting more labels\n",
+    "\n",
+    "In this example we used a modified version of the [Wall-Following Robot Navigation](https://www.openml.org/search?type=data&sort=runs&status=any&qualities.NumberOfClasses=gte_2&qualities.NumberOfInstances=between_1000_10000&id=1526) dataset, where we synthetically generated multiple annotations for each datapoint with some noise. "
```
Comment (jwmueller) [inline active_learning_multiannotator/active_learning.ipynb]:
```suggestion
    "This notebook implements the following steps:\n",
    "\n",
    "1. Establish consensus labels (for the already labeled data). Use them to train a classifier model and then obtain out-of-sample predicted probabilities for each example.\n", 
    "2. Compute active learning scores for every example, which estimate our current confidence in knowing its true label.\n",
    "3. Collect additional labels for the examples with the lowest active learning scores. These are the most potentially informative examples.\n",
    "4. Repeat the steps above to collect as many labels as your budget permits.\n",
"\n",
"The accuracy of the model trained on the resulting dataset will generally match that of the same model trained on a much larger set of randomly collected labels. I.e. this is the most cost-effective way to train an accurate classifier!\n",
    "\n",
    "In this example we use a modified version of the [Wall-Following Robot Navigation](https://www.openml.org/search?type=data&sort=runs&status=any&qualities.NumberOfClasses=gte_2&qualities.NumberOfInstances=between_1000_10000&id=1526) dataset, with multiple data annotators who each provide noisy labels."
```

File: active_learning_multiannotator/active_learning.ipynb [hunk 17]
```diff
@@ -0,0 +1,677 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We will load the following datafiles:\n",
+    "\n",
+    "- `multiannotator_labels` is a DataFrame that contains labels from multiple annotators for each datapoint\n",
+    "- `X_labeled` are the features for the examples that have been labeled by at least one annotator\n",
+    "- `X_unlabeled` are the features for the examples that have not been labeled by any annotator\n",
+    "\n",
+    "Additionally, we load these extra files for simulation and demontration purposed and would usually not be available:\n",
```
Comment (jwmueller) [inline active_learning_multiannotator/active_learning.ipynb:79]:
```suggestion
    "Additionally, we load these extra files for demonstration purposes and would usually not be available:\n",
```

File: active_learning_multiannotator/active_learning.ipynb [hunk 18]
```diff
@@ -0,0 +1,677 @@
...
+    "\n",
+    "- `extra_labels_labeled` and `extra_labels_unlabeled` are synthetically created extra labels that will be sampled from when we get more labels later in this notebook\n",
+    "- `X_test` and `true_labels_test` are used to measure the model accuracy "
```
Comment (jwmueller) [inline active_learning_multiannotator/active_learning.ipynb]:
```suggestion
    "We load the following datafiles:\n",
    "\n",
    "- `multiannotator_labels` is a DataFrame that contains labels from multiple annotators for each example (different number of annotations per example)\n",
    "- `X_labeled` are the features for the examples that have been labeled by at least one annotator\n",
    "- `X_unlabeled` are the features for the examples that have not yet been labeled by any annotator\n",
    "\n",
    "Additionally, we load some extra files for demo purposes, which will probably not be available in your application:\n",
    "\n",
    "- `extra_labels_labeled` and `extra_labels_unlabeled` are sampled from when we get ask annotators to provide additional labels later in this notebook\n",
    "- `X_test` and `true_labels_test` are used to measure the current model's predictive accuracy in each round of active learning"
```

File: active_learning_multiannotator/active_learning.ipynb [hunk 19]
```diff
@@ -0,0 +1,677 @@
...
+    "> `multiannotator_labels` should be a numpy array of pandas DataFrame where each column represents an annotator and each row represents an examples. The classes should be ingeters from 0 to num_classes - 1, where examples that were not labeled by a particular annotator are represented using `np.nan`. `multiannotator_labels` should also only contain the examples that have been labeled by at least one annotator (no row should consist of only `np.nan` values).\n",
+    "> \n",
+    "> `X_labeled` and `X_unlabeled` should be numpy arrays that have the same number of columns. However, if you do not have any unlabeled examples, you can also pass an empty array of `None` as the value of `X_unlabeled`.\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Demonstration of model improved after 15 rounds of active learning\n",
```
Comment (jwmueller) [inline active_learning_multiannotator/active_learning.ipynb]:
```suggestion
    "## Improving model accuracy over 15 rounds of active learning (collecting new labels) \n",
```

File: active_learning_multiannotator/active_learning.ipynb [hunk 20]
```diff
@@ -0,0 +1,677 @@
...
+    "\n",
+    "This demonstration runs an active learning loop for 15 rounds, choosing 100 examples to collect more labels for each round. Each round, we use the consensus labels of each labeled example to train a classifier (here we used an `ExtraTrees` classifier) and obtain out-of-sample predicted probabilites, which is then used to compute the active learning scores for every example. We then synthetically collect new labels (this process is meant to simulate getting a new annotator to annotated a selection of examples) and repeat the active learning loop. \n",
+    "\n",
+    "For the first round of training, we obtain consensus labels using majority vote as we do not have any predicted probabilites yet. However for subsequent rounds we can leverage the predicted probabilites for each example obtained from the previous round to use cleanlab's [CROWDLAB algorithm](https://docs.cleanlab.ai/stable/tutorials/multiannotator.html) to obtain higher quality consensus labels for the model.\n",
+    "\n",
+    "We also measure the model performance on a test set each round to demonstrate the improvement of the model."
```
Comment (jwmueller) [inline active_learning_multiannotator/active_learning.ipynb]:
```suggestion
    "[Optional step] We also measure the model performance on a test set each round to demonstrate the improvement of the model."
```

File: active_learning_multiannotator/active_learning.ipynb [hunk 21]
```diff
@@ -0,0 +1,677 @@
...
+    "    pred_probs, pred_probs_unlabeled = fit_predict_proba(\n",
+    "        ExtraTreesClassifier(),\n",
+    "        X_labeled,\n",
+    "        consensus_labels,\n",
+    "        cv_n_folds=5,\n",
+    "        X_unlabeled=X_unlabeled,\n",
+    "    )\n",
+    "\n",
+    "    # train a model on the full set of labeled data to get the model accuracy for the current round \n",
+    "    # (this is an optional step used for demonstration purposed in this notebook, in practical settings it is unlikely to have ground truth labels)\n",
```
Comment (jwmueller) [inline active_learning_multiannotator/active_learning.ipynb]:
```suggestion
    "    # train a model on the full set of labeled data to evaluate model accuracy for the current round \n",
    "    # this is an optional step for demonstration purposes, in practical applications you may not have ground truth labels\n",
```

File: active_learning_multiannotator/active_learning.ipynb [hunk 22]
```diff
@@ -0,0 +1,677 @@
...
+    "\n",
+    "    # get the indices of examples to collect more labels for\n",
+    "    relabel_idx, relabel_idx_unlabeled = get_idx_to_label(\n",
+    "        active_learning_score=active_learning_score,\n",
+    "        active_learning_score_unlabeled=active_learning_score_unlabeled,\n",
+    "        batch_size_to_label=batch_size_to_label,\n",
+    "    )\n",
+    "\n",
+    "    # format the data for the next round of active learning \n",
+    "    # (moving some unlabeled examples to the labeled pool because we are collecting labels for them)\n",
```
Comment (jwmueller) [inline active_learning_multiannotator/active_learning.ipynb]:
```suggestion
    "    # moving some unlabeled examples to the labeled pool because we are collecting labels for them\n",
```

File: active_learning_multiannotator/active_learning.ipynb [hunk 23]
```diff
@@ -0,0 +1,677 @@
...
+    "    (\n",
+    "        multiannotator_labels, relabel_idx_combined, X_labeled, X_unlabeled, pred_probs_labeled, \n",
+    "        pred_probs_unlabeled, extra_labels_labeled, extra_labels_unlabeled,\n",
+    "    ) = setup_next_iter_data(\n",
+    "        multiannotator_labels, relabel_idx, relabel_idx_unlabeled, X_labeled, X_unlabeled, pred_probs, \n",
+    "        pred_probs_unlabeled, extra_labels_labeled, extra_labels_unlabeled,\n",
+    "    )\n",
+    "\n",
+    "    # add a new annotator that provides new labels for the examples with lowest active learning scores (indices obtained above)\n",
```
Comment (jwmueller) [inline active_learning_multiannotator/active_learning.ipynb]: Note these long comments are getting cutoff for me in github's notebook viewer. So let's make sure all long comments are split into multiple lines and visible in the github viewer

File: active_learning_multiannotator/active_learning.ipynb [hunk 24]
```diff
@@ -0,0 +1,677 @@
...
+    "Below we show a simple example of how to obtain this information using the annotators' labels and the model predicted probabilites. Check out cleanlab's [multiannotator module](https://docs.cleanlab.ai/stable/tutorials/multiannotator.html) for more detailed information!"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# using the annotator labels and predicted probabilites to get the consensus labels and other statistics\n",
```
Comment (jwmueller) [inline active_learning_multiannotator/active_learning.ipynb]: for these one line commands, my personal pref is to move the comment above out of the code cell and have it be markdown text describing what the single line of code does.

File: active_learning_multiannotator/active_learning.ipynb (hunk 23 above)
Comment (huiwengoh) [reply active_learning_multiannotator/active_learning.ipynb]: All the comments look good in my github viewer (13 inch mac) now, let me know if something is still getting cut off in yours!

Comment (jwmueller) [reply active_learning_multiannotator/active_learning.ipynb]: yep lgtm now, not sure what happened

File: active_learning_multiannotator/utils/model_training.py [hunk 25]
```diff
@@ -0,0 +1,59 @@
...
+    get_num_classes,
+    append_extra_datapoint,
+    train_val_split,
+)
+
+
+def fit_predict_proba(
+    model,
+    X,
+    labels,
```
Comment (jwmueller) [inline active_learning_multiannotator/utils/model_training.py:20]: Probably want to clarify these are consensus labels, derived for a particular example by aggregating all collected  annotations into a single one. And the shape expected for `labels`

File: active_learning_multiannotator/utils/active_learning.py [hunk 26]
```diff
@@ -0,0 +1,92 @@
...
+    relabel_idx,
+    relabel_idx_unlabeled,
+    X,
+    X_unlabeled,
+    pred_probs,
+    pred_probs_unlabeled,
+    extra_labels=None,
+    extra_labels_unlabeled=None,
+):
+    """Updates inputs after additional labels have been collected in a single active learning round,
```
Comment (jwmueller) [inline active_learning_multiannotator/utils/active_learning.py]: throughout this notebook and code, let's try to refer to "active learning" as "multi-annotator active learning" or "active learning with multiple annotators", to more clearly distinguish our setup from the standard active learning one with a single label per example

File: active_learning_multiannotator/utils/active_learning.py [hunk 27]
```diff
@@ -0,0 +1,92 @@
+"""
+This file defines the helper functions for the active learning with multiple annotators simulation, which include data preperation
```
Comment (jwmueller) [inline active_learning_multiannotator/utils/active_learning.py]:
```suggestion
This file defines the helper functions for the active learning with multiple annotators simulation, which include data preparation
```

File: active_learning_multiannotator/active_learning.ipynb [hunk 28]
```diff
@@ -0,0 +1,721 @@
+{
```
Comment (jwmueller) [inline active_learning_multiannotator/active_learning.ipynb:1]: don't forget to stick some standard naming convention for the scores throughout the notebook and the helper files (in variable names, docstrings, text, etc). I believe we wanted to use name: "active learning quality scores"

**Review by jwmueller — APPROVED** [review]
I think the notebook is as simple as we can easily make it, without starting to cut useful information. Amazing work on this!
